clip_retrieval/clip_client.py

```diff
@@ -0,0 +1,137 @@
+        num_images: (optional) Number of images to return. Default is 40.
+        """
+        self.url = url
+        self.indice_name = indice_name
+        self.use_mclip = use_mclip
+        self.aesthetic_score = aesthetic_score
+        self.aesthetic_weight = aesthetic_weight
+        self.modality = modality.value
+        self.num_images = num_images
+
+    def query(
+        self,
+        text: str = None,
+        image: str = None,
+    ) -> List[Dict]:
+        """
+        Given text or image/s, search for other captions/images that are semantically similar.
+
+        Args:
+            text: text to be searched semantically.
+            image: base64 string of image to be searched semantically
+
+        Returns:
+            List of dictionaries containing the results in the form of:
+            [
+                {
+                    "id": 42,
+                    "similarity": 0.323424523424,
+                    "url": "https://example.com/image.jpg",
+                    "caption": "This is a caption",
+                },
+                ...
+            ]
+        """
+        if text and image:
+            raise ValueError("Only one of text or image can be provided.")
+        if text:
+            return self.__search_knn_api__(text=text)
+        elif image:
+            if image.startswith("http"):
+                return self.__search_knn_api__(image_url=image)
+            else:
+                assert Path(image).exists(), f"{image} does not exist."
+                return self.__search_knn_api__(image=image)
+        else:
+            raise ValueError("Either text or image must be provided.")
+
+    def __search_knn_api__(
+        self,
+        text: str = None,
+        image: str = None,
+        image_url: str = None,
+    ) -> List:
+        """
+        This function is used to send the request to the knn service.
+        It represents a direct API call and should not be called directly outside the package.
+
+        Args:
+            text: text to be searched semantically.
+            image: base64 string of image to be searched semantically.
+            image_url: url of the image to be searched semantically.
+
+        Returns:
+            List of dictionaries containing the results in the form of:
+            [
+                {
+                    "id": 42,
+                    "similarity": 0.323424523424,
+                    "url": "https://example.com/image.jpg",
+                    "caption": "This is a caption",
+                },
+                ...
+            ]
+
+        """
+        if image:
+            # Convert image to base64 string
+            with open(image, "rb") as image_file:
+                encoded_string = base64.b64encode(image_file.read())
+                image = str(encoded_string.decode("utf-8"))
+        return requests.post(
+            self.url,
+            data=json.dumps(
```

Comment:
I think that client is too opinionated

Having a lower level one that simply passes all the parameters/ expose all the features
Will be useful 

But let's do that later